index.js

```diff
@@ -169,6 +265,144 @@ class Pawn extends Piece {
+    this.hasMoved = true;
+  }
+}
+
+class Bishop extends Piece {
+  constructor(game, board, color, row, file) {
+    super(game, board, color);
+    this.row = row;
+    this.file = file;
+  }
+
+  name = "Bishop";
+
+  get moves() {
+    const available = [];
+
+    this.moveDiagonally(this.row, this.file, available, "up", "right");
+    this.moveDiagonally(this.row, this.file, available, "down", "left");
+    this.moveDiagonally(this.row, this.file, available, "up", "left");
+    this.moveDiagonally(this.row, this.file, available, "down", "right");
+    return available;
+  }
+
+  onMove(move) {
+    console.log(move);
+  }
+  get icon() {
+    return this.isWhite() ? "♗" : "♝";
+  }
+}
+
+class Queen extends Piece {
+  constructor(game, board, color, row, file) {
+    super(game, board, color);
+    this.row = row;
+    this.file = file;
+  }
+
+  name = "queen";
+
+  get moves() {
+    const available = [];
+    this.verticalMove(this.row, this.file, available, "up");
+    this.verticalMove(this.row, this.file, available, "down");
+    this.horizontalMove(this.row, this.file, available, "right");
+    this.horizontalMove(this.row, this.file, available, "left");
+
+    this.moveDiagonally(this.row, this.file, available, "up", "right");
+    this.moveDiagonally(this.row, this.file, available, "down", "left");
+    this.moveDiagonally(this.row, this.file, available, "up", "left");
+    this.moveDiagonally(this.row, this.file, available, "down", "right");
+    return available;
+  }
+
+  onMove(move) {
+    console.log(move);
+  }
+
+  get icon() {
+    return this.isWhite ? "♕" : "♛";
+  }
+}
+
+class King extends Piece {
+  constructor(game, board, color, row, file) {
+    super(game, board, color);
+    this.row = row;
+    this.file = file;
+    this.hasMoved = false;
+  }
+
+  name = "King";
+
+  moveOneSquare(row, file, arr, vert, horiz) {
+    const nextSquare = [row + vert, file + horiz];
+    if (
+      this.isValidSquare(...nextSquare) &&
+      !this.isSquareOccupied(...nextSquare)
+    ) {
+      arr.push(Move.fromSquare(nextSquare, this));
+    } else if (
+      this.isValidSquare(...nextSquare) &&
+      this.isSquareOccupied(...nextSquare) &&
+      this.getSquareContent(...nextSquare).color !== this.color
+    ) {
+      arr.push(
+        Move.fromSquare(nextSquare, this, this.isSquareOccupied(...nextSquare))
+      );
+    }
+    return arr;
+  }
+
+  get moves() {
+    const available = [];
+
+    this.moveOneSquare(this.row, this.file, available, 1, 0);
+    this.moveOneSquare(this.row, this.file, available, -1, 0);
+    this.moveOneSquare(this.row, this.file, available, 0, 1);
+    this.moveOneSquare(this.row, this.file, available, 0, -1);
+    this.moveOneSquare(this.row, this.file, available, 1, 1);
+    this.moveOneSquare(this.row, this.file, available, 1, -1);
+    this.moveOneSquare(this.row, this.file, available, -1, 1);
+    this.moveOneSquare(this.row, this.file, available, -1, -1);
+    return available;
+  }
```

Comment:
See if you can add an optional third parameter to our `getLegalDirectionalMoves` function that should be a `magnitude: number` that will be whatever distance we want to look for available moves. If the consumer of the function does not provide this optional third parameter, we should default it to `Board.LANE_SIZE - 1`.
If you do this right you should be able to make this look like this:
```suggestion
  get moves() {
    const directions = [[-1, -1], [-1, 1], [1, -1], [1, 1], [0, -1], [-1, 0], [0, 1], [1, 0]]
    const available = directions.flatMap(dir => getLegalDirectionalMoves(this, direction))
    return available
  }
  }
```